spec/mem/tree_borrows/memory.md

```diff
@@ -85,19 +101,48 @@ impl<T: Target> TreeBorrowsMemory<T> {
             return ret(ptr);
         };
 
+        let freeze_info = ptr_type.safe_pointee().map(|pointee_info| pointee_info.freeze);
+        let pointee_nonfreeze_bytes = freeze_info.map(|strategy| {
+            match strategy {
+                UnsafeCellStrategy::Sized { bytes } => bytes,
+                // TODO: Implement for unsized cases
+                _ => List::new(),
+            }
+        });
+        let is_freeze = freeze_info.map_or(true, |strategy| strategy.is_freeze());
+        let protected = new_perm.protected;
+
         let child_path = self.mem.allocations.mutate_at(alloc_id.0, |allocation| {
+            let size = allocation.size();
+
+            let location_states: List<LocationState> = if pointee_nonfreeze_bytes.is_some() {
+                (Int::ZERO..size.bytes()).map(|i| {
+                    let i = Size::from_bytes(i).unwrap();
+                    // Check if `i` is included in any of the ranges. O(size * |pointee_nonfreeze_bytes|)
+                    // TODO: More performant way to do this?
+                    let frozen = !pointee_nonfreeze_bytes.unwrap().any(|(start, end)| start <= i && i < end);
+                    let perm = if frozen { new_perm.freeze_perm } else { new_perm.nonfreeze_perm };
+                    LocationState {
+                        accessed: Accessed::No, // This gets updated to `Accessed::Yes` if `allocation.extra.root.access` runs
+                        permission: perm
+                    }
+                }).collect()
+            } else {
+                LocationState::new_list(new_perm.freeze_perm, size)
+            };
+
             // Create the new child node
             let child_node = Node {
                 children: List::new(),
-                location_states: LocationState::new_list(permission, allocation.size()),
+                location_states,
                 protected,
             };
 
             // Add the new node to the tree
             let child_path = allocation.extra.root.add_node(parent_path, child_node);
 
-            // If this is a non-zero-sized reborrow, perform read on the new child, updating all nodes accordingly.
-            if pointee_size.bytes() > 0 {
+            // If this is a non-zero-sized reborrow, perform read on the new child if needed, updating all nodes accordingly.
+            if pointee_size.bytes() > 0 && if is_freeze {new_perm.freeze_access} else {new_perm.nonfreeze_access} {
```

Comment:
This currently reads the entire range. But this might read the nonfreeze bytes when it should not, which is wrong. It needs to be more finely-grained.